### PR TITLE
Use Zipline in the Emoji Search sample

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt [hunk 1]
```diff
@@ -98,17 +100,19 @@ internal class OutboundCall(
...
-      val callbackName = endpoint.generateName()
-      val callback = RealSuspendCallback(callbackName, continuation, serializer)
-      endpoint.set<SuspendCallback>(callbackName, EmptySerializersModule, callback)
-      endpoint.outboundChannel.invokeSuspending(
-        instanceName,
-        funName,
-        arguments.toTypedArray(),
-        callbackName
-      )
+      CoroutineScope(context.endpoint.dispatcher).launch {
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt:105]: I'm thinking that you're looking for this pattern: https://github.com/square/retrofit/blob/6c34d7b6b534325408948d37c573e1fe40154a2a/retrofit/src/main/java/retrofit2/KotlinExtensions.kt#L113-L114

A naked `CoroutineScope` is a huge red flag for me.

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt [hunk 2]
```diff
@@ -77,7 +76,7 @@ class Endpoint internal constructor(
       callbackName: String
     ) {
       val handler = inboundHandlers[instanceName] ?: error("no handler for $instanceName")
-      CoroutineScope(EmptyCoroutineContext).launch(dispatcher) {
+      CoroutineScope(dispatcher).launch {
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt:79]: Very skeptical of both the before and after here, once again regarding naked scopes.

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+class EmojiSearchZipline {
+  private val executorService = Executors.newSingleThreadExecutor { Thread(it, "Zipline") }
+  private val dispatcher = executorService.asCoroutineDispatcher()
+  private val zipline = Zipline.create(dispatcher)
+  private val hostApi = RealHostApi()
+
+  fun produceModelsIn(
+    coroutineScope: CoroutineScope,
+    eventFlow: MutableSharedFlow<EmojiSearchEvent>,
```
Comment (JakeWharton) [inline samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt]:
```suggestion
    eventFlow: Flow<EmojiSearchEvent>,
```

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+      val eventsFlowReference = eventFlow.asFlowReference(EmojiSearchEvent.serializer())
+      val modelsFlowReference = presenter.produceModels(eventsFlowReference)
+
+      val modelsFlow = modelsFlowReference.get(EmojiSearchViewModel.serializer())
+      modelsStateFlow.emitAll(modelsFlow)
+    }
+
+    job.invokeOnCompletion {
+      dispatcher.dispatch(EmptyCoroutineContext) { zipline.close() }
```
Comment (JakeWharton) [inline samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt:61]: Use executor directly? You already use it below, although dispatcher shutdown is actually a forthcoming API so maybe you won't in the future...

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/RealHostApi.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import okhttp3.Callback
+import okhttp3.Headers.Companion.toHeaders
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+
+class RealHostApi : HostApi {
+  private val client = OkHttpClient()
+  override suspend fun httpCall(url: String, headers: Map<String, String>): String {
+    return suspendCoroutine { continuation ->
```
Comment (JakeWharton) [inline samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/RealHostApi.kt]: Should really be `suspendCancelableCoroutine` for correctness.

**Review by JakeWharton — APPROVED** [review]

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt (hunk 1 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt:105]: Agreed! I’d like some help with this! I tried the above, it didn’t work:

```
java.lang.ClassCastException: class kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to class kotlinx.coroutines.internal.DispatchedContinuation (kotlin.coroutines.jvm.internal.CompletedContinuation and kotlinx.coroutines.internal.DispatchedContinuation are in unnamed module of loader 'app')
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:107)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at app.cash.zipline.internal.bridge.OutboundCall$RealSuspendCallback.call(outbound.kt:129)
	at app.cash.zipline.internal.bridge.OutboundCall$invokeSuspending$2$1$run$1$create$1.call(outbound.kt)
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt (hunk 2 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/Endpoint.kt:79]: I’m trying and failing to post work to the dispatcher. Best way to do that?

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt (hunk 4 above)
Comment (swankjesse) [reply samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt:61]: Both work for me.

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt (hunk 1 above)
Comment (JakeWharton) [reply zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/outbound.kt:105]: Well to be honest the Retrofit code is stolen from Roman so... we can figure it out the right thing eventually.

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/RealHostApi.kt (hunk 5 above)
Comment (swankjesse) [reply samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/RealHostApi.kt]: Fixed.

Comment (swankjesse) [reply samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/RealHostApi.kt]: *suspendCancellableCoroutine

**Pull request merged**